In [82]:
import nltk
from nltk.corpus import stopwords
import re
!pip install .whl file
!pip install fasttext
import fasttext
import numpy as np
import pandas as pd

ERROR: Invalid requirement: '.whl'
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [83]:
from google.colab import files


uploaded = files.upload()


Saving Review1.txt to Review1 (2).txt
Saving Review2.txt to Review2 (2).txt
Saving Review3.txt to Review3 (2).txt


# TASK1

In [84]:
Review1= open('Review1.txt','r').read() 
Review2= open('Review2.txt','r').read() 

In [85]:
challenge_train = Review1+"\n"+Review2

In [86]:
with open ('challenge_train.txt', 'w') as fp: #Saving the merge file
    fp.write(challenge_train)

# TASK2

In [87]:
nltk.download('stopwords')
Stopwords = stopwords.words('english')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [88]:
line= open('challenge_train.txt','r').read() 
line = re.sub(r'[^\w\s]', '', line) # removing punctuation 
words = line.split()
for r in words:
    if not r in Stopwords:
        appendFile = open('cl_challenge_train.txt','a')
        appendFile.write(" "+r)
        appendFile.close()

## TASK3

In [89]:
line= open('Review3.txt','r').read() 
line = re.sub(r'[^\w\s]', '', line) # removing punctuation 
words = line.split()
for r in words:
    if not r in Stopwords:
        appendFile = open('cleaned_Review3.txt','a')
        appendFile.write(" "+r)
        appendFile.close()

## TASK4

In [90]:
model_sup = fasttext.train_supervised(input='cl_challenge_train.txt', lr=0.7, epoch=50, wordNgrams=3, bucket=20000)

## TASK5

In [91]:
test_num, precision, recall = model_sup.test('cleaned_Review3.txt')
f_score = (2*precision*recall)/(precision+recall)
f_score

0.0010368066355624676

## TASK6

In [92]:
check = []
for k in np.arange(1,10,1):
  for t in np.arange(0.1,1,0.1):
    test_num, precision, recall = model_sup.test('cleaned_Review3.txt', k=k, threshold=t)
    f_score = (2*precision*recall)/(precision+recall)
    check.append(f_score)

list1 = []
list2 = []

for k in np.arange(1,10,1):
  for t in np.arange(0.1,1,0.1):
    list1.append(k) 
    list2.append(t)
    
mydf = pd.DataFrame([list1,list2,check]).transpose().sort_values(2,ascending = False)
mydf.columns = ['K Value','Threshold value','f_score']
display(mydf.dropna())

K Value  Threshold value   f_score
38      5.0              0.3  0.002073
30      4.0              0.4  0.002073
56      7.0              0.3  0.002073
55      7.0              0.2  0.002073
54      7.0              0.1  0.002073
63      8.0              0.1  0.002073
48      6.0              0.4  0.002073
47      6.0              0.3  0.002073
46      6.0              0.2  0.002073
45      6.0              0.1  0.002073
64      8.0              0.2  0.002073
39      5.0              0.4  0.002073
65      8.0              0.3  0.002073
37      5.0              0.2  0.002073
36      5.0              0.1  0.002073
66      8.0              0.4  0.002073
29      4.0              0.3  0.002073
72      9.0              0.1  0.002073
12      2.0              0.4  0.002073
75      9.0              0.4  0.002073
74      9.0              0.3  0.002073
73      9.0              0.2  0.002073
9       2.0              0.1  0.002073
10      2.0              0.2  0.002073
11      2.0              0.3  0.002073
57      7.0              0.4  0.002073
28      4.0              0.2  0.002073
18      3.0              0.1  0.002073
19      3.0              0.2  0.002073
20      3.0              0.3  0.002073
21      3.0              0.4  0.002073
27      4.0              0.1  0.002073
67      8.0              0.5  0.001037
58      7.0              0.5  0.001037
0       1.0              0.1  0.001037
49      6.0              0.5  0.001037
40      5.0              0.5  0.001037
1       1.0              0.2  0.001037
31      4.0              0.5  0.001037
22      3.0              0.5  0.001037
13      2.0              0.5  0.001037
4       1.0              0.5  0.001037
3       1.0              0.4  0.001037
2       1.0              0.3  0.001037
76      9.0              0.5  0.001037

## As we can see highest f_score: 0.002073 we are getting with K-Value: 5	  and Threshold-value: 0.3